picard/util/pipe.py

```diff
@@ -0,0 +1,359 @@
+# -*- coding: utf-8 -*-
+#
+# Picard, the next-generation MusicBrainz tagger
+#
+# Copyright (C) 2022 skelly37
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+from abc import (
+    ABCMeta,
+    abstractmethod,
+)
+import concurrent.futures
+import os
+from tempfile import NamedTemporaryFile
+from typing import (
+    List,
+    Optional,
+)
+
+from picard import (
+    PICARD_APP_ID,
+    log,
+)
+from picard.const.sys import (
+    IS_MACOS,
+    IS_WIN,
+)
+
+
+if IS_WIN:
+    import win32pipe  # type: ignore
+    import win32file  # type: ignore
+    from pywintypes import error as WinApiError  # type: ignore
+
+
+class PipeError(Exception):
+    MESSAGE = None
+
+    def __init__(self, *messages):
+        if self.MESSAGE:
+            self.messages = (self.MESSAGE,) + tuple(messages)
+        else:
+            self.messages = tuple(messages)
+
+    def __str__(self):
+        messages_str = "\n  ".join(str(m) for m in self.messages)
+        if not messages_str:
+            messages_str = "unknown"
+        return f"ERROR: {messages_str}"
+
+
+class PipeErrorInvalidArgs(PipeError):
+    MESSAGE = "Pipe() args argument has to be iterable"
+
+
+class PipeErrorInvalidAppData(PipeError):
+    MESSAGE = "Pipe() app_name and app_version arguments have to be str"
+
+
+class PipeErrorNotFound(PipeError):
+    MESSAGE = "Pipe doesn't exist"
+
+
+class PipeErrorBroken(PipeError):
+    MESSAGE = "Pipe is broken"
+
+
+class PipeErrorInvalidResponse(PipeError):
+    MESSAGE = "Invalid response from pipe"
+
+
+class PipeErrorWin(PipeError):
+    MESSAGE = "Windows API error"
+
+
+class PipeErrorNoPermission(PipeError):
+    MESSAGE = "No permissions for creating a pipe"
+
+
+class PipeErrorNoDestination(PipeError):
+    MESSAGE = "No available dirs to place a pipe"
+
+
+class AbstractPipe(metaclass=ABCMeta):
+    NO_RESPONSE_MESSAGE: str = "No response from FIFO"
+    MESSAGE_TO_IGNORE: str = '\0'
+    TIMEOUT_SECS: float = 1.5
+
+    @classmethod
+    @property
+    @abstractmethod
+    def PIPE_DIRS(cls):
+        raise NotImplementedError
```

Comment:
@skelly37 @zas I like this approach. Overall the use of platform specific classes has made the code much easier to understand and better separated.

Any specific reason why `@abstractproperty` was not used here?

```python
    @abstractproperty
    def readonly(self):
        return self.__x
```